python/tank/bootstrap/async_bootstrap.py

```diff
@@ -0,0 +1,278 @@
+        """
+
+        # Make the QThread object exit its event loop.
+        self._thread.quit()
+
+        # Make the worker operate with the main thread affinity
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        if self._failed_callback:
+            # Handle cleanup after failed completion of the toolkit bootstrap.
+            self._failed_callback(self._toolkit_manager.TOOLKIT_BOOTSTRAP_PHASE, exception)
+
+
+class _BootstrapToolkitWorker(QtCore.QObject):
+    """
+    Bootstrap worker that can bootstrap an :class:`~sgtk.Sgtk` instance asynchronously in a background thread.
+
+    This class defines 3 Qt signals:
+    - ``progressing`` with signature ``(float, str, int, int)``
+      emitted while the bootstrap toolkit worker is progressing in its work in the background,
+      where ``float, str, int, int``are the current progress value, the progress message to report,
+      the current item number being looped over and the maximum item number being looped over.
+    - ``completed`` with signature ``(Sgtk)``
+      emitted when the bootstrap toolkit worker successfully completes its work in the background,
+      where ``Sgtk`` is the bootstrapped :class:`~sgtk.Sgtk` instance.
+    - ``failed`` with signature ``(Exception)``
+      emitted when the bootstrap toolkit worker fails to complete its work in the background,
+      where ``Exception`` is the python exception raised while bootstrapping.
+    """
+
+    # Qt signal emitted while the bootstrap toolkit worker is progressing in its work in the background.
+    progressing = QtCore.Signal(float, str, int, int)
+
+    # Qt signal emitted when the bootstrap toolkit worker successfully completes its work in the background.
+    completed = QtCore.Signal(Sgtk)
+
+    # Qt signal emitted when the bootstrap toolkit worker fails to complete its work in the background.
+    failed = QtCore.Signal(Exception)
+
+    def __init__(self, toolkit_manager, engine_name, entity):
+        """
+        Initializes an instance of the bootstrap toolkit worker.
+
+        :param toolkit_manager: :class:`~sgtk.bootstrap.ToolkitManager` instance bootstrapping the engine.
+        :param engine_name: Name of the engine used to resolve a configuration.
+        :param entity: Shotgun entity used to resolve a project context.
+        """
+
+        super(_BootstrapToolkitWorker, self).__init__()
+
+        self._toolkit_manager = toolkit_manager
+        self._engine_name = engine_name
+        self._entity = entity
+
+    @QtCore.Slot()
+    def work(self):
+        """
+        Bootstraps a toolkit instance for the given engine and entity and
+        signal the progress, and the completion or the failure of this work.
+        """
+
+        # Preserve the original toolkit manager progress reporting callback.
+        original_progress_callback = self._toolkit_manager._progress_cb
```

Comment:
Here the protected member is used to retrieve the current progress callback but below there's a public method to set it. Should it just be a property with both get and set? 

BTW, sincerest apologies if this has been discussed already and I missed it. Feel free to ignore any of my gibberish.